dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/DefaultTrackerImportService.java

```diff
@@ -78,144 +68,76 @@ public class DefaultTrackerImportService implements TrackerImportService {
 
   @Nonnull private final TrackerUserService trackerUserService;
 
-  @Nonnull private final Notifier notifier;
-
+  /* Import is not meant to be annotated with @Transactional.
+   * PreHeat and Commit phases are separated transactions, other
+   * phases do not need to be in a transaction. */
   @Override
-  public ImportReport importTracker(TrackerImportParams params) {
+  public ImportReport importTracker(
+      TrackerImportParams params, TrackerObjects trackerObjects, JobProgress jobProgress) {
     User user = trackerUserService.getUser(params.getUserId());
-    params.setUser(user);
-
-    TimingsStats opsTimer = new TimingsStats();
-
-    startImport(params);
-
-    try {
-      TrackerBundle trackerBundle = preHeat(params, opsTimer);
-
-      Map<TrackerType, Integer> bundleSize = calculatePayloadSize(trackerBundle);
-
-      preProcess(opsTimer, trackerBundle);
-
-      ValidationReport validationReport = validate(params, opsTimer, trackerBundle);
 
-      if (exitOnError(validationReport, params)) {
-        return buildReportAndNotify(params, validationReport, opsTimer, bundleSize);
-      }
+    jobProgress.startingStage("Running PreHeat");
+    TrackerBundle trackerBundle =
+        jobProgress.runStage(() -> trackerBundleService.create(params, trackerObjects, user));
 
-      PersistenceReport persistenceReport = commit(params, opsTimer, trackerBundle);
+    Map<TrackerType, Integer> bundleSize = calculatePayloadSize(trackerBundle);
```

Comment:
I tend to put even such fast and quite safe steps in a stage just in case there is a exception happening it gets handled more gracefully. But it will also work if you don't do that.

Reply:
Done